token/services/db/sql/driver.go

```diff
@@ -88,29 +89,18 @@ func (d *DBOpener) compileOpts(cp core.ConfigProvider, tmsID token.TMSID) (*Opts
 	return opts, nil
 }
 
-func (d *DBOpener) OpenSQLDB(driverName, dataSourceName string, maxOpenConns int, skipPragmas bool) (*sql.DB, error) {
-	logger.Infof("connecting to [%s] database", driverName) // dataSource can contain a password
+type dbKey struct {
+	driverName, dataSourceName string
+}
 
-	id := driverName + dataSourceName
-	var p *sql.DB
-	d.mutex.RLock()
-	p, ok := d.dbs[id]
-	if ok {
-		logger.Infof("reuse [%s] database (cached)", driverName)
-		d.mutex.RUnlock()
-		return p, nil
-	}
-	d.mutex.RUnlock()
+func key(k dbKey) string {
+	return k.driverName + k.dataSourceName
+}
 
-	d.mutex.Lock()
-	defer d.mutex.Unlock()
+func (d *DBOpener) OpenSQLDB(driverName, dataSourceName string, maxOpenConns int, skipPragmas bool) (*sql.DB, error) {
+	logger.Infof("connecting to [%s] database", driverName) // dataSource can contain a password
 
-	// check again
-	p, ok = d.dbs[id]
-	if ok {
-		return p, nil
-	}
-	p, err := sql.Open(driverName, dataSourceName)
+	p, err := d.dbCache.Get(dbKey{driverName: driverName, dataSourceName: dataSourceName})
```

Comment:
if an error occurs after the creation of a new entry in the cache, then the cache is not cleaned. Should the cash be cleaned in case of an error? Another option is to move the rest of the code in the function that create objects in the cache.
In the previous code, the map was not updated.